File: cf_sdk/src/main/java/com/example/cf_sdk/logic/changebankapi/source/remote/AuthenticationRemoteDatasource.java
package com.example.cf_sdk.logic.changebankapi.source.remote;

import com.example.cf_sdk.logic.changebankapi.Endpoints;
import com.example.cf_sdk.logic.changebankapi.network.api.ApiConfig;
import com.example.cf_sdk.logic.changebankapi.network.api.AuthenticationApi;
import com.example.cf_sdk.logic.changebankapi.source.AuthenticationRepository;
import com.example.cf_sdk.logic.changebanksdk.model.FingerprintResponse;
import com.example.cf_sdk.logic.changebanksdk.model.Session;
import com.example.cf_sdk.logic.changebanksdk.model.member.SecurityProfile;
import com.example.cf_sdk.logic.changebanksdk.parameter.Parameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.CheckPasscodeEnabledParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.GetSecurityProfileParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.LoginParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.ResendTwoFactorParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.SecurityProfileParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.TwoFactorParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.UpdateCredentialsParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.UpdatePasswordParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.authentication.UsernameParameters;
import com.example.cf_sdk.logic.changebanksdk.response.ChangebankResponse;
import com.example.cf_sdk.logic.changebanksdk.source.AuthenticationDatasource;
import com.google.common.base.Optional;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;

/**
 *
 * Remote datasource to call authentication endpoints.
 */

public class AuthenticationRemoteDatasource implements AuthenticationDatasource {
    private static final String TAG = AuthenticationRepository.class.getSimpleName();
    private AuthenticationApi mAuthenticationApi;

    @Inject
    public AuthenticationRemoteDatasource(AuthenticationApi authenticationApi) {
        mAuthenticationApi = authenticationApi;
    }

    @Override
    public Single<Session> login(LoginParameters loginParameters) {
        return mAuthenticationApi.login(loginParameters.getHeaders().get(ApiConfig.USER_URL) + Endpoints.Auth.AUTHENTICATION, loginParameters.getHeaders(), loginParameters);
    }

    @Override
    public Single<ChangebankResponse> updateCredentials(UpdateCredentialsParameters updateCredentialsParameters) {
        return mAuthenticationApi.updateCredentials(updateCredentialsParameters.getHeaders(), updateCredentialsParameters.getMemberId(), updateCredentialsParameters);
    }

    @Override
    public Single<Session> getSession() {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support getSession");
    }

    @Override
    public Single<LoginParameters.LoginMethodType> getLoginMethodType() {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support getLoginMethodType");
    }

    @Override
    public Single<Optional<byte[]>> getPinSalt() {
        throw new UnsupportedOperationException("AuthenticationCacheDatasource not support getPinSalt");
    }

    @Override
    public Single<ChangebankResponse> checkUsernameExists(UsernameParameters usernameParameters) {
        return mAuthenticationApi.checkUsernameExists(usernameParameters.getHeaders(), usernameParameters);
    }

    @Override
    public Single<Optional<String>> getSavedUsername() {
        throw new UnsupportedOperationException("AuthenticationCacheDatasource not support getSavedUsername");
    }

    @Override
    public Single<Optional<String>> getCachedUsername() {
        throw new UnsupportedOperationException("AuthenticationCacheDatasource not support getCachedUsername");
    }

    @Override
    public Single<ChangebankResponse> logout(Parameters parameters) {
        return mAuthenticationApi.logout(parameters.getHeaders());
    }

    @Override
    public Single<ChangebankResponse> changePassword(UpdatePasswordParameters updatePasswordParameters) {
        return mAuthenticationApi.changePassword(updatePasswordParameters.getHeaders(), updatePasswordParameters);
    }

    @Override
    public Single<Boolean> checkPasscodeEnabled(CheckPasscodeEnabledParameters parameters) {
        throw new UnsupportedOperationException("AuthenticationCacheDatasource not support checkPasscodeEnabled");
    }

    @Override
    public void saveSession(Session session) {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support saveSession");
    }

    @Override
    public void savePinSalt(byte[] pinSalt, String username) {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support savePinSalt");
    }

    @Override
    public void saveUsername(String username) {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support saveUsername");
    }

    @Override
    public void clearAllAuthenticationDatasourceCache() {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support clearAllAuthenticationDatasourceCache");
    }

    @Override
    public void clearPinAndFingerprint() {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support clearPinAndFingerprint");
    }

    @Override
    public void updateLocalStorage(LoginParameters loginParameters) {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support updateLocalStorage");
    }

    @Override
    public Observable<FingerprintResponse> readFingerprint(String username) {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support readFingerprint");
    }

    @Override
    public Single<FingerprintResponse> readFingerprintInstant(String username) {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support readFingerprintInstant");
    }

    @Override
    public Completable writeFingerprint(String username, String passwordHash) {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support writeFingerprint");
    }

    @Override
    public Single<Boolean> canStoreFingerprintSecurely() {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support canStoreFingerprintSecurely");
    }

    @Override
    public Completable clearAllFingerprint() {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support clearAllFingerprint");
    }

    @Override
    public Single<Boolean> isUserFingerprintSaved(String username) {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support isUserFingerprintSaved");
    }

    @Override
    public Completable clearSpecificFingerprint(String username) {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support clearSpecificFingerprint");
    }

    @Override
    public Single<Session> resendLoginMFA(ResendTwoFactorParameters resendTwoFactorParameters) {
        return mAuthenticationApi.resendCode(resendTwoFactorParameters.getHeaders(), resendTwoFactorParameters.getMemberId());
    }

    @Override
    public Single<Session> verifyLoginMFA(TwoFactorParameters twoFactorParameters) {
        return mAuthenticationApi.twoFactor(twoFactorParameters.getHeaders(), twoFactorParameters);
    }

    @Override
    public Single<ChangebankResponse> updateSecurityProfile(SecurityProfileParameters securityProfileParameters) {
        return mAuthenticationApi.updateSecurityProfile(securityProfileParameters.getHeaders(), securityProfileParameters.getMemberId(), securityProfileParameters);
    }

    @Override
    public Single<SecurityProfile> getSecurityProfile(GetSecurityProfileParameters getSecurityProfileParameters) {
        return mAuthenticationApi.getSecurityProfile(getSecurityProfileParameters.getHeaders(), getSecurityProfileParameters.getMemberId());
    }

    @Override
    public Completable clearPasscode() {
        throw new UnsupportedOperationException("AuthenticationRemoteDatasource not support clearPasscode");
    }
}

File: cf_sdk/src/main/java/com/example/cf_sdk/changebankapi/source/remote/AuthenticationRemoteDatasource.kt
package com.example.cf_sdk.changebankapi.source.remote;


import com.example.cf_sdk.changebankapi.model.FingerprintResponse
import com.example.cf_sdk.defination.response.Session
import com.example.cf_sdk.changebankapi.model.member.SecurityProfile
import com.example.cf_sdk.changebankapi.network.api.AuthenticationApi
import com.example.cf_sdk.changebankapi.parameter.Parameters
import com.example.cf_sdk.changebankapi.parameter.authentication.GetSecurityProfileParameters
import com.example.cf_sdk.changebankapi.parameter.authentication.LoginMethodType
import com.example.cf_sdk.changebankapi.parameter.authentication.LoginParameters
import com.example.cf_sdk.changebankapi.parameter.authentication.ResendTwoFactorParameters
import com.example.cf_sdk.changebankapi.parameter.authentication.SecurityProfileParameters
import com.example.cf_sdk.changebankapi.parameter.authentication.TwoFactorParameters
import com.example.cf_sdk.changebankapi.parameter.authentication.UsernameParameters
import com.example.cf_sdk.defination.response.ChangebankResponse
import com.example.cf_sdk.changebankapi.source.AuthenticationRepository
import com.example.cf_sdk.defination.request.AccessTokenParameter


import com.google.common.base.Optional
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single

/**
 *
 * Remote datasource to call authentication endpoints.
 */
class AuthenticationRemoteDatasource(private val mAuthenticationApi: AuthenticationApi) :
    AuthenticationDatasource {


    override val session: Single<Session?>?
        get() {
            throw UnsupportedOperationException("AuthenticationRemoteDatasource not support getSession")
        }
    override val loginMethodType: Single<LoginMethodType?>?
        get() {
            throw UnsupportedOperationException("AuthenticationRemoteDatasource not support getLoginMethodType")
        }
    override val pinSalt: Single<Optional<ByteArray?>?>?
        get() {
            throw UnsupportedOperationException("AuthenticationCacheDatasource not support getPinSalt")
        }

    override fun checkUsernameExists(usernameParameters: UsernameParameters?): Single<ChangebankResponse?>? {
        return mAuthenticationApi.checkUsernameExists(
            usernameParameters!!.headers,
            usernameParameters
        )
    }

    override val savedUsername: Single<Optional<String?>?>?
        get() {
            throw UnsupportedOperationException("AuthenticationCacheDatasource not support getSavedUsername")
        }
    override val cachedUsername: Single<Optional<String?>?>?
        get() {
            throw UnsupportedOperationException("AuthenticationCacheDatasource not support getCachedUsername")
        }

    override fun logout(parameters: Parameters?): Single<ChangebankResponse?>? {
        return mAuthenticationApi.logout(parameters!!.headers)
    }


    override fun saveSession(session: Session?) {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support saveSession")
    }

    override fun savePinSalt(pinSalt: ByteArray?, username: String?) {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support savePinSalt")
    }

    override fun saveUsername(username: String?) {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support saveUsername")
    }

    override fun clearAllAuthenticationDatasourceCache() {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support clearAllAuthenticationDatasourceCache")
    }

    override fun clearPinAndFingerprint() {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support clearPinAndFingerprint")
    }

    override fun updateLocalStorage(loginParameters: LoginParameters?) {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support updateLocalStorage")
    }

    override fun readFingerprint(username: String?): Observable<FingerprintResponse?>? {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support readFingerprint")
    }

    override fun readFingerprintInstant(username: String?): Single<FingerprintResponse?>? {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support readFingerprintInstant")
    }

    override fun writeFingerprint(username: String?, passwordHash: String?): Completable? {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support writeFingerprint")
    }

    override fun canStoreFingerprintSecurely(): Single<Boolean?>? {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support canStoreFingerprintSecurely")
    }

    override fun clearAllFingerprint(): Completable? {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support clearAllFingerprint")
    }

    override fun isUserFingerprintSaved(username: String?): Single<Boolean?>? {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support isUserFingerprintSaved")
    }

    override fun clearSpecificFingerprint(username: String?): Completable? {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support clearSpecificFingerprint")
    }

    override fun resendLoginMFA(resendTwoFactorParameters: ResendTwoFactorParameters?): Single<Session?>? {
        return mAuthenticationApi.resendCode(
            resendTwoFactorParameters!!.headers,
            resendTwoFactorParameters.memberId
        )
    }

    override fun verifyLoginMFA(twoFactorParameters: TwoFactorParameters?): Single<Session?>? {
        return mAuthenticationApi.twoFactor(twoFactorParameters!!.headers, twoFactorParameters)
    }

    override fun updateSecurityProfile(securityProfileParameters: SecurityProfileParameters?): Single<ChangebankResponse?>? {
        return mAuthenticationApi.updateSecurityProfile(
            securityProfileParameters!!.headers,
            securityProfileParameters.memberId,
            securityProfileParameters
        )
    }

    override fun getAccessToken(accessTokenParameter: AccessTokenParameter?): Single<Session?>? {
        // todo: Replace url it with new one
        return mAuthenticationApi.getAccessToken("http://e49.vtxn-qadev.cf-cloud.net:8080/v1/auth/token",accessTokenParameter?.headers,accessTokenParameter)
    }


    override fun getSecurityProfile(getSecurityProfileParameters: GetSecurityProfileParameters?): Single<SecurityProfile?>? {
        return mAuthenticationApi.getSecurityProfile(
            getSecurityProfileParameters!!.headers,
            getSecurityProfileParameters.memberId
        )
    }

    override fun clearPasscode(): Completable? {
        throw UnsupportedOperationException("AuthenticationRemoteDatasource not support clearPasscode")
    }

    companion object {
        private val TAG = AuthenticationRepository::class.java.simpleName
    }
}